### PR TITLE
fix roctracer references in hip and cleanup code in rocprofiler-dev recipe

### PR DESCRIPTION
**Pull request by srekolam**
with out the /opt/rocm installtion we are currently seeing  warning WARNING "Profiling API header not found. Disabling roctracer integration. Use -DPROF_API_HEADER_PATH=<path to prof_protocol.h header>" while building  hip.
Also , when the rocm rpm's are installed in the /opt/rocm path it is picking up the header files from /opt/rocm path which i think is not correct ( -I/opt/rocm/roctracer/include/ext is included )   while compiling the hip files eg hip_context.cpp  

the change inside the hip recipe is meant to fix this issue . After the changes i do not see the warning and also header file is properly picked.
example o/p from the terminal - Profiling API: /tmp/root/spack-stage/spack-stage-hip-4.1.0-csml555jn4u7qbl3mgplbxrjvw36gvlp/spack-src/roctracer/inc/ext 

the changes to the rocprofiler-dev is cleanup code. basically replaces with smaller lines of code.

testing - i have built hip for rocm-4.2.0 ,rocm-4.1.0 and rocfft recipe for rocm-4.2.0 and rocm-4.1.0 
for rocprofiler-dev i have built and installed for rocm-4.2.0 and rocm-4.1.0 releases.

### CODE REVIEW

**Comment by haampie** [comment]
Can you maybe use `roctracer-dev` as a separate package instead of a resource?

Ugh, I see there's a circular dependency, great!

**Comment by haampie** [comment]
also it doesn't build for me:

```
[ 34%] Built target rocprof-link
[ 36%] Linking C executable c_test
cd /tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-build-rqkv3pj/test && /opt/spack/linux-ubuntu20.04-zen2/gcc-10.2.0/cmake-3.20.5-suunrvsam7gab4li77k2dw4vjy5zhowc/bin/cmake -E cmake_link_script CMakeFiles/c_test.dir/link.txt --verbose=1
/lib/spack/env/gcc/gcc -O3 -DNDEBUG -rdynamic CMakeFiles/c_test.dir/app/c_test.c.o -o c_test 
make[2]: Leaving directory '/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-build-rqkv3pj'
[ 36%] Built target c_test
In file included from /tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.h:29,
                 from /tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.cpp:23:
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:81:16: error: 'string' in namespace 'std' does not name a type
   81 |   virtual std::string Name() const = 0;
      |                ^~~~~~
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:28:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
   27 | #include <stdint.h>
  +++ |+#include <string>
   28 | #include <map>
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:84:16: error: 'string' in namespace 'std' does not name a type
   84 |   virtual std::string SymbName() { return Name() + ".kd"; }
      |                ^~~~~~
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:84:11: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
   84 |   virtual std::string SymbName() { return Name() + ".kd"; }
      |           ^~~
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h: In member function 'bool TestKernel::SetHostDescr(const uint32_t&, const TestKernel::des_id_t&, const uint32_t&)':
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:108:19: error: 'malloc' was not declared in this scope
  108 |       descr.ptr = malloc(size);
      |                   ^~~~~~
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/ctrl/test_kernel.h:28:1: note: 'malloc' is defined in header '<cstdlib>'; did you forget to '#include <cstdlib>'?
   27 | #include <stdint.h>
  +++ |+#include <cstdlib>
   28 | #include <map>
In file included from /tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.cpp:23:
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.h: At global scope:
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.h:50:8: error: 'string' in namespace 'std' does not name a type
   50 |   std::string Name() const { return std::string("SimpleConvolution"); }
      |        ^~~~~~
/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-src/test/simple_convolution/simple_convolution.h:30:1: note: 'std::string' is defined in header '<string>'; did you forget to '#include <string>'?
   29 | #include "ctrl/test_kernel.h"
  +++ |+#include <string>
   30 | 
make[2]: *** [test/CMakeFiles/ctrl.dir/build.make:135: test/CMakeFiles/ctrl.dir/simple_convolution/simple_convolution.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/tmp/spack-stage/spack-stage-rocprofiler-dev-4.2.0-rqkv3pj2nvni3zgkn56e3juksrrq37eq/spack-build-rqkv3pj'
make[1]: *** [CMakeFiles/Makefile2:327: test/CMakeFiles/ctrl.dir/all] Error 2
```

maybe disable these tests?

**Comment by haampie** [comment]
For hip you can just add a `depends_on('roctracer-dev')` instead of a resource, that'd be much better.

**Comment by srekolam** [comment]
@haampie  when i just add the depends_on('roctracer-dev@' + ver, when='@' + ver) while doing
spack install -v hip i get the following error - 
Error: KeyError: 'No spec with name hip-rocclr in hip@4.2.0%gcc@8.4.1~ipo build_type=RelWithDebInfo patches=2a4190477b7d9206b9cd8d70770ba0bc007273cbe54772efb12f9ca2e37c0392,e276c4acf3d37712b6bea306fea34f539d3c4f743471e9da208b5eb17b16ae67 arch=linux-centos8-skylake_avx512'

/spack/var/spack/repos/builtin/packages/hip/package.py:230, in flag_handler:
        228    def flag_handler(self, name, flags):
        229        if name == 'cxxflags' and '@3.7.0:' in self.spec:
  >>    230            incl = self.spec['hip-rocclr'].prefix.include
        231            flags.append('-I {0}/compiler/lib/include'.format(incl))
        232            flags.append('-I {0}/elf'.format(incl))

Also, after removing  depends_on('py-setuptools@:44', type='build') inside roc-tracer-dev and while doing spack install roctracer-dev i do get the following 

Error: KeyError: 'No spec with name hip in roctracer-dev@4.2.0%gcc@8.4.1~ipo build_type=RelWithDebInfo arch=linux-centos8-skylake_avx512'

/spack/var/spack/repos/builtin/packages/roctracer-dev/package.py:47, in setup_build_environment:
         45    def setup_build_environment(self, build_env):
         46        spec = self.spec
  >>     47        build_env.set("HIP_PATH", spec['hip'].prefix)

i am thinking this could be due to circular dependency .. inside roctracer-dev , we already have  depends_on(hip) and now adding depends_on(roctracer-dev) will show up some issues. hence added as resource.

File: var/spack/repos/builtin/packages/hip/package.py [hunk 1]
```diff
@@ -234,6 +255,8 @@ def flag_handler(self, name, flags):
 
     def cmake_args(self):
         args = [
+            '-DPROF_API_HEADER_PATH={0}/roctracer/inc/ext'.format(
+                self.stage.source_path),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/hip/package.py]:
```suggestion
            self.define('PROF_API_HEADER_PATH', join_path(self.stage.source_path, 'roctracer', 'inc', 'ext'))
```

**Comment by haampie** [comment]
Hm, gotta love circular dependencies :D. Another approach is to define a separate package `rocprofiler-api` that just installs / copies the headers but does not build the library (so, don't invoke cmake, just do `install(path/to/headers, prefix/include/...)`). What do you think about that?

Generally we should limit use of resource, in particular cause it makes it difficult to apply patches.

File: var/spack/repos/builtin/packages/hip/package.py (hunk 1 above)
Comment (srekolam) [reply var/spack/repos/builtin/packages/hip/package.py]: okay ..sure..will do the changes.

**Comment by srekolam** [comment]
> Hm, gotta love circular dependencies :D. Another approach is to define a separate package `rocprofiler-api` that just installs / copies the headers but does not build the library (so, don't invoke cmake, just do `install(path/to/headers, prefix/include/...)`). What do you think about that?
> 
> Generally we should limit use of resource, in particular cause it makes it difficult to apply patches.

Okay..sure..do you have any pointers about any package that is currently doing it ..i can take it as reference and will do it  for next release..? let me know if this okay ?

File: var/spack/repos/builtin/packages/rocprofiler-dev-api/package.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class RocprofilerDevApi(Package):
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocprofiler-dev-api/package.py]: You're not doing anything with rocprofiler during install, can you please turn this into roctracer-dev-api?

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/rocprofiler-dev-api/package.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            url='https://github.com/ROCm-Developer-Tools/roctracer/archive/rocm-{0}.tar.gz'.format(d_version),
+            sha256=d_shasum,
+            expand=True,
+            destination='',
+            placement='roctracer',
+            when='@{0}'.format(d_version)
+        )
+
+    def install(self, spec, prefix):
+        print("Installing  ... ")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocprofiler-dev-api/package.py]:
```suggestion
```

**Comment by haampie** [comment]
I've pushed two changes to your branch @srekolam, can you try these? This is what I meant.

**Comment by srekolam** [comment]
> I've pushed two changes to your branch @srekolam, can you try these? This is what I meant.

I have verified it..built the hip package for rocm-4.2.0 as well as the rocprofiler-dev package. looks good to me.

**Review by haampie — APPROVED** [review]

**Pull request merged**